### PR TITLE
maven@3.6 3.6.3 (new formula)

### PR DESCRIPTION
**Pull request by fredcooke**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [ ] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

Have not checked locally, GH actions can do that for me. Commits contain prose, not 1980s vintage formatting. Learned from bad changes in prior PR #86188 and advice from @SMillerDev on that PR (thanks!). Confirmed 3.1+ are not end of life, but all 3.0.X and older are deprecated and entirely unsupported. Maven is a tool that you tend to build your stack around and stick with for a long time. Our stack is pretty good and can work with 3.8.1 and 3.6.X but current 3.8.2 is broken for us (and 3.5, 3.3, 3.2 that are already included in homebrew core are too old for us).

### CODE REVIEW

**Comment by fredcooke** [comment]
As an aside, there was a 3 month gap between maven 3.6.2 and 3.6.3 and like 3.8.2 3.6.2 had some serious faults (though I don't think those ones in particular affected us). 

2019-08-27 -> 2019-11-25

Brew would have had broken maven or old out of date maven, only, for many users during that 3 month period. I don't think in the case of Maven that the rules serve the users very well, however fortunately 3.6 should be accepted and that's good enough for us :-)

**Comment by SMillerDev** [comment]
Does maven actually specify when 3.6 will be EOL?

**Comment by SMillerDev** [comment]
Because the way I read https://maven.apache.org/docs/history.html is that they do not support 3.6 because there is a newer version of the 3.1+ releases.

**Comment by carlocab** [comment]
> Because the way I read https://maven.apache.org/docs/history.html is that they do not support 3.6 because there is a newer version of the 3.1+ releases.

This is my reading as well, especially since 3.8.1 [fixes two CVEs](https://maven.apache.org/docs/3.8.1/release-notes.html) with no sign of it being backported to 3.6.x.

**Comment by fredcooke** [comment]
By that logic you should immediately remove 3.5, 3.3, and 3.2 - but I'm sure they're useful to someone or the "not enough downloads" thing would have ripped them out, right? 3.6 is more valid than all of those IMO, having a slew of fixes over and above 3.5 and so on down to 3.2 which I haven't used in quite a few years.

Nevertheless, I've tried to do as the build says and add a bottle do block, but I failed to use the instructions from https://docs.brew.sh/Bottles to get anything useful happening on my work Mac. I got it to build from `install --build-bottle` but it failed to bottle up using `bottle` claiming it "wasn't installed from" what it was installed from etc. 

Any help getting that bottle sorted out would be great :-)

**Comment by carlocab** [comment]
> By that logic you should immediately remove 3.5, 3.3, and 3.2

Not exactly. We're much more demanding of new formulae than we are of existing ones. In any case, each of those formulae is deprecated, and will be disabled and deleted eventually.

**Comment by fredcooke** [comment]
Well, aside from the changes required to make this a compliant version specific formula it's the exact one lifted from a pio

**Comment by carlocab** [comment]
>So it's not new, it's reincarnated

That's worse: we have [even higher standards](https://docs.brew.sh/Versions#acceptable-versioned-formulae) for reincarnated formulae than we do for any other new one.

**Comment by fredcooke** [comment]
Then how did 3.2 and 3.3 and 3.5 get in? They're dinosaur tech :-D 3.6.3 is only just replaced a few months ago with 3.8.1 (good release) which has only just been replaced a few weeks ago with 3.8.2 (bad release) and due to brew policy everyone is stuck with that bad 3.8.2 on the basis that it may be more secure in real world use (probably not) when it's actually not functional at all in various circumstances. Being more secure is no use at all if it simply does not work :-D On the face of it the brew polices make a lot of sense, but in this case they make very little sense. 

Right up until 3.8.1 was released everyone that brew installed maven got 3.6.3 and so forth, so if it had lower uptake it was only due to brew itself getting lower uptake of maven users during that period. And if the uptake went down then that was due to auto upgrades away to 3.8.1 which was okay (for us, but maybe not all) and then again to 3.8.2 (broken for us, maybe okay for others). I feel like it should at least have some blue/green ie, keep the last one you rolled as well as the current one, but it's not structured that way and the rules say no 3.8.1 (which is the best option right now) and that we should be able to have 3.6 (acceptable at least for a while).

I think it would be flat-out ridiculous if you allowed 3.2 but not 3.6 - that literally makes zero sense at all. Allowing both makes good sense. Allowing 3.8.1 is against the rules in the "even higher standards" (even stricter, not higher) but would make more sense than 3.6 which in turn makes more sense than 3.5 and so on down the list. 

Nearly 3am, I'm out for the count. Please reconsider, there's about 50 people with broken 3.8.2 in our sets of teams alone right now and rolling out instructions on using a custom tap to so many people across so many organisations is not as easy as you might think. Much better if it "just works". With 3.6 in place we can change the tooling to depend on that specifically and it'll all fall into line. Zzzzz time.

**Comment by fredcooke** [comment]
I've dropped the maven crew a note about this PR so hopefully someone with some Authority and even more than my 14 years experience using it can chime in and add their 2c for against this motion :-)

**Comment by carlocab** [comment]
They were added years ago -- things were different back then. That they weren't deprecated and then removed a while ago was an oversight that does not justify making another mistake here.

> rolling out instructions on using a custom tap to so many people across so many organisations is not as easy as you might think.

Perhaps not, but telling people to
```
brew install your-tap-user/your-tap-name/your-maven-formula
```
isn't really that much harder than telling people to `brew install maven@3.6`, which you would need to do in the event that this PR were accepted.

File: Formula/maven@3.6.rb [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+      (bin/basename).write_env_script file, Language::Java.overridable_java_home_env
+    end
+  end
+
+  test do
+    (testpath/"pom.xml").write <<~EOS
+      <?xml version="1.0" encoding="UTF-8"?>
+      <project xmlns="https://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+        xsi:schemaLocation="https://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
```
Comment (michael-o) [inline Formula/maven@3.6.rb:43]: The scheme has been superseded many years ago.

**Comment by michael-o** [comment]
The general approach is that every new version supersedes old version. There is no old minor version support. There is only two reasons two use an old version:
* You are happy with
* You are confined with an old version due to regulatory or technical reasons

We request to verify a bug in the most recent version. No excuses. There are two ways to have a fix in an old version:
* You can convince a committer to solve it
* Incentive through money

I will release 3.8.3 on Monday. Forget anything before 3.5.4 or even 3.6.3. If 3.8.3 proves to be regression  free, forget 3.6.3 as well.

**Comment by michael-o** [comment]
Upshot: Don't waste your time for too old version.

**Comment by jonchang** [comment]
Thanks for your feedback @michael-o, and thanks for contributing to Homebrew, @fredcooke. I think we won't add this formula since your use-case seems to be solved in the forthcoming 3.8.3, and either way you can use `brew tap-new` to create `maven@3.6`. Please see [How to Create and Maintain a Tap](https://docs.brew.sh/How-to-Create-and-Maintain-a-Tap) for more details. 

Homebrew is maintained by unpaid volunteers and we have to draw the line somewhere, and that includes aggressively pruning the amount of old versions we need to maintain. It's fine that you're using it for your clients, but your business needs are superseded by the needs of Homebrew's maintainers.

File: Formula/maven@3.6.rb (hunk 1 above)
Comment (rfscholte) [reply Formula/maven@3.6.rb:43]: It shuld be `xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/xsd/maven-4.0.0.xsd"`

**Comment by michael-o** [comment]
We are all unpaid volunteers too.

File: Formula/maven@3.6.rb (hunk 1 above)
Comment (fredcooke) [reply Formula/maven@3.6.rb:43]: Sure, I just pulled it from Brew history as is and made the tweaks to get it to build (more) cleanly against the very strict rules.

**Comment by fredcooke** [comment]
> Perhaps not, but telling people to
> 
> ```
> brew install your-tap-user/your-tap-name/your-maven-formula
> ```
> 
> isn't really that much harder than telling people to `brew install maven@3.6`, which you would need to do in the event that this PR were accepted.

Correct, looking to not do either, looking to make maven@3.6 the dependency of an existing tap that every single one of them has, so that it happens automatically and seamlessly and they get a stable build experience. I guess the thing to do here is to pull this file into that existing tap side by side with it, or maybe the 3.8.1 version, and depend on it that way. That solves it with zero effort/communications and produces a "just works" system, too. 

Thanks to all of the unpaid volunteers in both Brew and Maven who chimed in here :-) appreciated. I'm not here as an unpaid volunteer today and do not use MacOS by choice, but I've been hacking on code for 31 years, using Linux since early 2002 or maybe earlier, and maven since 2.0.4 or maybe earlier. I've contributed to heaps of stuff and run a few small open sauce projects of my own, too. I get it :-D 

My final comment on this PR is to suggest a refactor and different model of no-support-if-not-latest and slightly more nested tree, maintainers, please consider this to better support the needs of real world software and its users:

change from `<formula>.rb`
change to:
`<formula>/latest` -> symlink to `./1.2.4` (most recent version)
`<formula>/1.2.4` (most recent version)
`<formula>/1.2.3` (previous patch version)
SNIP <all prior patches of current minor>
`<formula>/1.2.1` (before previous patch version)
`<formula>/1.2.0` (first release of current minor version)
`<formula>/1.1.7` (ultimate patch to previous minor version)
`<formula>/1.0.4` (ultimate patch to one before previous minor version)
etc and cull from the back end of that based on stats

Reality is, all software has bugs and being able to back grade chunks of it is regularly useful in EVERY ecosystem.

Maven central holds jars/poms/etc back 15+ years, you can still pull all of it.
Debian has volunteer maintained repos back a decade, you can still maintain most of them
etc. But brew: I want last week's version. NO :-D this does NOT make sense for any serious software delivery system. I hope you can see that and seriously think about a system that provides, but not supports, older stuff, for people to make their OWN calls about, rather than having it stuffed down their throats the Apple take-it-or-leave-it way :-)

Cheers. Looking forward to 3.8.3 :-) The issue in question was this one: https://issues.apache.org/jira/browse/MINVOKER-283

**Comment by fredcooke** [comment]
Edit because it hid my "html tag" on the SNIP line which wasn't quoted rather than just escaping/displaying it, sigh.

**Comment by SMillerDev** [comment]
> I hope you can see that and seriously think about a system that provides, but not supports, older stuff, for people to make their OWN calls about, rather than having it stuffed down their throats the Apple take-it-or-leave-it way.

We have exactly that with `brew extract` and custom taps. You can take any older version of the formula from brew, we don't support that older formula. But you can do that in a tap.

Great to hear your bug will be gone on Monday though.

**Pull request closed without merging**